gax/src/main/java/com/google/api/gax/tracing/OpencensusTracerFactory.java

```diff
@@ -0,0 +1,115 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.InternalApi;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link ApiTracerFactory} to build instances of {@link OpencensusTracer}.
+ *
+ * <p>This class wraps the {@link Tracer} provided by Opencensus in {@code Tracing.getTracer()}. It
+ * will be used to create new spans and wrap them in {@link OpencensusTracer} defined in gax.
+ */
+@InternalApi("For google-cloud-java client use only")
+public final class OpencensusTracerFactory implements ApiTracerFactory {
+  @Nonnull private final Tracer internalTracer;
+  @Nullable private final String clientNameOverride;
+
+  /**
+   * Instantiates a new instance capturing the {@link io.opencensus.trace.Tracer} in {@code
+   * Tracing.getTracer}.
+   */
+  public OpencensusTracerFactory() {
+    this(null);
+  }
+
+  /**
+   * Instantiates a new instance capturing the {@link io.opencensus.trace.Tracer} in {@code
+   * Tracing.getTracer}. It will also override the service name of the grpc stub with a custom
+   * client name. This is useful disambiguate spans created outer manual written wrappers and around
+   * generated gapic spans.
+   *
+   * @param clientNameOverride the client name that will override all of the spans' client name.
+   */
+  public OpencensusTracerFactory(@Nullable String clientNameOverride) {
+    this(Tracing.getTracer(), clientNameOverride);
+  }
+
+  /**
+   * Instantiates a new instance with an explicit {@link io.opencensus.trace.Tracer}. It will also
+   * override the service name of the grpc stub with a custom client name. This is useful
+   * disambiguate spans created outer manual written wrappers and around generated gapic spans.
+   *
+   * @param internalTracer the Opencensus tracer to wrap.
+   * @param clientNameOverride the client name that will override all of the spans' client name.
+   */
+  @InternalApi("Visible for testing")
+  OpencensusTracerFactory(Tracer internalTracer, @Nullable String clientNameOverride) {
+    this.internalTracer =
+        Preconditions.checkNotNull(internalTracer, "internalTracer can't be null");
```

Comment:
Stylistic: I think in the rest of gax we usually do not put messages in `checkNotNull()` in such cases (as it is clear what the problem is without any messages, the exception will have line number to see which arg is wrong).

Reply:
I find it easier to debug if the arg name is in the error message. The issue will be at the call site level, forcing the developer to look at the callee implementation to see which of their args is null and then debug at the outer scope seems unfriendly. But if you have strong feelings I can remove it.